pageserver/src/layered_repository/inmemory_layer.rs

```diff
@@ -0,0 +1,407 @@
+//!
+//! An in-memory layer stores recently received page versions in memory. The page versions
+//! are held in a BTreeMap, and there's another BTreeMap to track the size of the relation.
+//!
+
+use crate::layered_repository::storage_layer::Layer;
+use crate::layered_repository::storage_layer::PageVersion;
+use crate::layered_repository::SnapshotLayer;
+use crate::repository::{BufferTag, RelTag, WALRecord};
+use crate::walredo::WalRedoManager;
+use crate::PageServerConf;
+use crate::ZTimelineId;
+use anyhow::{bail, Result};
+use bytes::Bytes;
+use log::*;
+use std::collections::BTreeMap;
+use std::ops::Bound::Included;
+use std::sync::Mutex;
+
+use zenith_utils::lsn::Lsn;
+
+static ZERO_PAGE: Bytes = Bytes::from_static(&[0u8; 8192]);
+
+pub struct InMemoryLayer {
+    conf: &'static PageServerConf,
+    timelineid: ZTimelineId,
+    tag: RelTag,
+
+    ///
+    /// This layer contains all the changes from 'start_lsn'. The
+    /// start is inclusive. There is no end LSN; we only use in-memory
+    /// layer at the end of a timeline.
+    ///
+    start_lsn: Lsn,
+
+    // FIXME: the three mutex-protected fields below should probably be protected
+    // by a single mutex.
+
+    /// If this relation was dropped, remember when that happened. Lsn(0) means
+    /// it hasn't been dropped
+    drop_lsn: Mutex<Lsn>,
```

Comment:
I wonder why do we use three mutexes here for different fields rather than extract all shared data i one structure and use one mutex for synchronizing access to it.